### DISCUSSION TITLE
OpenTelemetry integration

### DISCUSSION THREAD

**Discussion by jack-pappas** [grafana/pyroscope#1162] [Ideas]
@omarabid mentioned in #1077 a new implementation of tagging that'll work across multiple threads. Has there been any consideration towards having that work with OpenTelemetry-instrumented code? A good number of libraries are now instrumented with [OpenTelemetry](https://opentelemetry.io/), and it'd be useful to be able to cross-reference collected OpenTelemetry data against Pyroscope; I think that would achieve the same goal as the new cross-thread tagging support.

If Pyroscope doesn't have a way to reference the OpenTelemetry data directly (e.g. by having the UI reach out to fetch tracing span data and/or metrics from some storage backend), it should still be possible to make this work by implementing an OpenTelemetry sink which writes the span data to Pyroscope. That'd let users of OpenTelemetry-instrumented libraries just configure their app to use the Pyroscope sink, without needing to add any separate Pyroscope-specific instrumentation.

**Comment by Rperry2174** [comment]
Great point @jack-pappas ! We actually have an integration like this for Go already with our [otelpyroscope](https://github.com/pyroscope-io/otelpyroscope) package! I think that it will be slightly more complicated to add this for Java, but definitely within reason... 

We also have a [proof-of-concept](https://github.com/pyroscope-io/pyroscope/tree/main/examples/tracing/jaeger) in our examples folder of linking Pyroscope to Jaeger (with a custom Jaeger UI).

Im curious, what tracing tool are you using that you'd like to link with profiles?

**Comment by diktamxx** [comment]
Has there been any progress on this issue? Does the current version support collecting profiler data via the OTEL protocol?

**Comment by kinthaiofficial** [comment]
Pyroscope + OpenTelemetry integration is particularly valuable for agent workloads where the bottleneck isn't always obvious from traces alone — you need profiling to see which functions are actually burning CPU/memory within a traced span.

For agent systems, the combination enables:

**Correlating trace with profile** — when an LLM call span takes 8 seconds, is it waiting on network (visible in trace), or is there CPU burn on tokenization/sampling (visible in profile)? Without the correlation, you're guessing.

**Per-agent profiling** — if you propagate agent context through the profile labels (Pyroscope's `WithContext` + custom labels), you can break down CPU/memory by `agent_id`. An agent that's consuming 40% of CPU while doing 10% of the work is a signal worth investigating.

**Memory profiling for context growth** — agent context windows grow over time. A heap profile correlated with the conversation ID shows exactly when and where memory is allocated for context, which is the first step to optimizing context compaction.

**Inference server profiling** — for local model serving (llama.cpp, vLLM, TGI), profiling the inference server during agent load shows whether the bottleneck is GPU compute, memory bandwidth, or CPU preprocessing. Trace tells you it's slow; profile tells you why.

OTel exemplars are the technical mechanism for trace-profile correlation — spans can carry a reference to the profile snapshot taken during that span. Grafana Cloud supports this natively; self-hosted needs some wiring.

We've been using continuous profiling for KinthAI's agent inference layer to catch expensive context operations: https://blog.kinthai.ai/openclaw-multi-tenancy-why-vm-per-user-doesnt-scale

Is the OTel integration primarily for trace-profile correlation, or for automatic profiling trigger based on trace signals?
